### PR TITLE
pi-ui POC & css module introduction

### PR DESCRIPTION
**Pull request by amassarwi**
We have decided to try integrate [pi-ui UI library](https://github.com/decred/pi-ui) at this stage, which should improve the molecularity and help us keep a consistent look & feel across decred's different products.

The following `pi-ui` related changes were done:
 -  use `ThemeProvider` to specify the themes & fonts (deleted `themes.less`).
 -  update getting started & settings page to toggle theme using pi-ui's useTheme hook.
 -   use `pi-ui`'s `Button` & `Link` in proposals list & proposal details pages.

The following changes can be done next in pi-ui's context:
 - delete icons.less and provide it using pi-ui's themes object (should be minor, see `styles/themes`)
 - use status bar component for voting progress  
 - use radio button component for voting preference
 - use pi-ui's tabs for governance page tabs 

This diff includes some work towards #2439 by introducing css modules - proposal details page was migrated & proposals list page was partly migrated, we should continue with incremental approach here.
It also includes work from #2438 - hooks polish in proposal details page/components, custom hooks, usage of `useMemo` & `useCallback` hooks, separated files for all helper functional components.

The rest of the governance pages can be refactored next to use pi-ui's components, css modules, hooks & functional components.

### CODE REVIEW

File: app/components/views/LNPage/index.js [hunk 1]
```diff
@@ -27,4 +27,5 @@ const LNActivePage = () =>
 const LNPage = ({ lnActive }) =>
   lnActive ? <LNActivePage /> : <LNConnectPage />;
 
+
```
Comment (vctt94) [inline app/components/views/LNPage/index.js:30]: unecessary change

File: app/components/shared/ExternalLink.js [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
 const clicker = (isTestNet, href, hrefTestNet) => () => {
   const url = hrefTestNet && isTestNet ? hrefTestNet : href;
   shell.openExternal(url);
 };
 
-const ExternalLink = ({ className, href, children, hrefTestNet, isTestNet }) => (
-  <a className={className} onClick={clicker(isTestNet, href, hrefTestNet) } href="#">
-    {children}
-  </a>
+const ExternalLink = ({ className, href, children, hrefTestNet, isTestNet, isButton }) => (
```
Comment (vctt94) [inline app/components/shared/ExternalLink.js:10]: I believe we can have only one tag here. No need to have the option to this component being a button or a link.

Just make sure it work in all places we use it (it should I believe)

Comment (amassarwi) [reply app/components/shared/ExternalLink.js:10]: It needed here in order to supper normal links (e.g the example in the following screenshot) - we could make use pi-ui's link here as well tho - I will give a try.

image

Comment (tiagoalvesdulce) [reply app/components/shared/ExternalLink.js:10]: Here I'll suggest using the Link component from pi-ui. It does have a customComponent property where you can pass the Button component e.g `<Link href="#" customComponent={() => <Button>Click me</Button>} />`. I added it to the docs here: https://github.com/decred/pi-ui/pull/250

File: app/components/shared/ExternalLink.js [hunk 3]
```diff
@@ -1,15 +1,20 @@
...
+  isButton ? (
+    <Button size="sm" className={className} onClick={clicker(isTestNet, href, hrefTestNet) } href="#">
```
Comment (tiagoalvesdulce) [inline app/components/shared/ExternalLink.js:12]: It doesn't make sense to pass an href prop to the Button component.

File: app/components/shared/ExternalLink.js (hunk 2 above)
Comment (tiagoalvesdulce) [reply app/components/shared/ExternalLink.js:10]: With the above comment, you can probably remove the `isButton` flag. I'm not a big fan of the `isSomething` flag when it's not state related. I'm making an effort to remove them from pgui as well. You could use the same behavior as the Link component and pass a render component to ExternalLink instead. if the component is used multiple times you can create a ExternalLinkAsButton component and it will be basically `<ExternalLink renderComponent={() => <Button ... />}/>`. IMO the code becomes more declarative this way.

File: app/components/views/ProposalDetails/ChooseVoteOption.jsx [hunk 4]
```diff
@@ -76,11 +80,12 @@ function ChooseVoteOption({
   });
 
   const error = state && state.context && getError(state.context.error);
+  // TODO: use memo here! currently this is being recaclulated on each new render!
```
Comment (tiagoalvesdulce) [inline app/components/views/ProposalDetails/ChooseVoteOption.jsx:83]: 👍

File: app/components/views/ProposalDetails/ProposalDetails.jsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      if (!hasEligibleTickets) {
+        return <NoElligibleTicketsVotingInfo {...{ showPurchaseTicketsPage }} />;
+      }
+      return <ChooseVoteOption {...{
+        viewedProposalDetails, voteOptions, setVoteOption, newVoteChoice,
+        eligibleTicketCount, currentVoteChoice, votingComplete: false
+      }} />;
+    }
+    return <ProposalNotVoting />;
+  }, [ proposalStatus, voteStatus, hasTickets, hasEligibleTickets, currentVoteChoice, viewedProposalDetails, eligibleTicketCount, newVoteChoice, setVoteOption, showPurchaseTicketsPage, voteOptions ]);
```
Comment (tiagoalvesdulce) [inline app/components/views/ProposalDetails/ProposalDetails.jsx:49]: Isn't this line too long? Also, good job memoizing this but I'd rather split it into another component and use the `React.memo()` instead. This component is quite polluted by JSX already.

File: app/style/Icons.less [hunk 6]
```diff
@@ -2,6 +2,9 @@
 @icon-root: '../style/icons';
 @animation-root: '../style/animations';
 
+
+// TODO: load these as theme variables!
```
Comment (tiagoalvesdulce) [inline app/style/Icons.less:6]: 👍

**Comment by tiagoalvesdulce** [comment]
The order of my comments was messed up somehow. You can send me a DM on matrix if things are unclear.

**Comment by amassarwi** [comment]
@vctt94 @tiagoalvesdulce thanks for the review guys, will be addressed.

File: app/components/views/ProposalDetails/ProposalDetails.jsx (hunk 5 above)
Comment (amassarwi) [reply app/components/views/ProposalDetails/ProposalDetails.jsx:49]: Good point mate, I've considered splitting the component here, thought maybe doing it in another PR but actually it make sense to do it at this stage when the css module are there.
I can give it a shot.

**Comment by matheusd** [comment]
My only comment is to be careful in changing the components visuals since it affects the UI as designed by @linnutee and co. Probably best to include some screenshots of the wip.

**Comment by amassarwi** [comment]
@matheusd Sure, we have the specs in mind.
Actually this PR was suppose to include the new eligible tickets components provided here #2428.
We decided to split it and add another PR with the new component.

File: .github/workflows/nodejs.yml [hunk 7]
```diff
@@ -15,7 +15,9 @@ jobs:
       with:
         node-version: ${{ matrix.node-version }}
 
-    - name: yarn
+    - name: yarn install 1.21.1
```
Comment (amassarwi) [inline .github/workflows/nodejs.yml:18]: This isn't valid and will reverted, was just a try to resolve the build issue!

**Comment by amassarwi** [comment]
> Right now it is possible to change vote option in a finished proposal.

@vctt94: I was able to reproduce the same behavior on master!
I will include the fix for that as well as part of this PR.

File: app/components/shared/ExternalLink.js (hunk 2 above)
Comment (vctt94) [reply app/components/shared/ExternalLink.js:10]: Great. I am not really a fan of overcomplicating components as well. Glad you pointed the directions so amass can deal with it :)

File: app/components/views/GovernancePage/Proposals/index.js [hunk 8]
```diff
@@ -54,15 +54,16 @@ function Proposals() {
...
 
   // TODO move reducers which only control local states from reducer/governance.js
   // to here.
   const [ tab, setTab ] = useReducer(() => getProposalsTab(location));
 
-  const dispatch = useDispatch();
   useEffect(() => {
     dispatch(setLastPoliteiaAccessTime());
-  }, []);
+  }, [ dispatch ]);
```
Comment (vctt94) [inline app/components/views/GovernancePage/Proposals/index.js:65]: Why you add this dispatch here? I left it empty to be used as a "componentDidMount"

File: app/components/layout/TabbedPage.js [hunk 9]
```diff
@@ -103,7 +103,7 @@ function TabbedPage(props) {
   const [ dir, setDir ] = useState("l2r");
   useEffect( () => {
     setMatchedTab(getMatchedTab(location, props.children));
-  }, []);
+  }, [ props.children, location ]);
```
Comment (vctt94) [inline app/components/layout/TabbedPage.js:106]: same here

File: app/components/layout/TabbedPage.js [hunk 10]
```diff
@@ -116,7 +116,7 @@ function TabbedPage(props) {
       : "l2r";
     setDir(dir);
     setMatchedTab(matchedTab);
-  }, [ location ]);
+  }, [ location, previous, props ]);
```
Comment (vctt94) [inline app/components/layout/TabbedPage.js:119]: I don't think we need previous here, as if previous values are changing, it means that current values changed

File: app/components/views/GovernancePage/Proposals/index.js (hunk 8 above)
Comment (tiagoalvesdulce) [reply app/components/views/GovernancePage/Proposals/index.js:65]: It's safer and obeys the `exhaustive-deps` rule. You can read more about it [here](https://github.com/facebook/react/issues/14920). The dispatch function is created outside the component context so it shouldn't change. It should work exactly the same.

File: app/components/layout/TabbedPage.js (hunk 9 above)
Comment (tiagoalvesdulce) [reply app/components/layout/TabbedPage.js:106]: Here is more case dependant, if the deps break the code and you have a really good case for keeping the dep list empty it's worth disabling the lint rule inline with `// eslint-disable-next-line react-hooks/exhaustive-deps`. Sometimes is misleading disabling it, tho. Take a look at the links in [this comment](https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528).

File: app/components/shared/ExternalLink.js (hunk 2 above)
Comment (vctt94) [reply app/components/shared/ExternalLink.js:10]: @amassarwi can you add tiago's comment?

Comment (amassarwi) [reply app/components/shared/ExternalLink.js:10]: Sure, will do!

File: app/components/views/GovernancePage/Proposals/index.js (hunk 8 above)
Comment (amassarwi) [reply app/components/views/GovernancePage/Proposals/index.js:65]: @tiagoalvesdulce thanks for reading my mind 👍

File: app/components/layout/TabbedPage.js (hunk 9 above)
Comment (amassarwi) [reply app/components/layout/TabbedPage.js:106]: I think we shouldn't disable the rule, I think we either should specify array with *all* deps, or with no array => will run one time like componentdidmount AFAIR.

Comment (amassarwi) [reply app/components/layout/TabbedPage.js:106]: Here are some insights why we should be strict with the deps rule everywhere:
https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies
https://reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often

TLDR: we should break the useEffect to several ones instead of omitting deps, see second link above for more insights.

Comment (amassarwi) [reply app/components/layout/TabbedPage.js:106]: Also, we shouldn't use deep properties like `props.children` in the deps array, instead we should always define another const outside the useEffect and use it instead inside the `useEffect` and in the deps array!

Comment (tiagoalvesdulce) [reply app/components/layout/TabbedPage.js:106]: Yeah, the useEffect hook should be used for a specific task and you can have multiple useEffect hooks. Even better if you have custom hooks with descriptive names for each of these tasks if they are related or just too long.

File: app/components/views/GovernancePage/Proposals/index.js (hunk 8 above)
Comment (tiagoalvesdulce) [reply app/components/views/GovernancePage/Proposals/index.js:65]: 😂😂😂

**Comment by amassarwi** [comment]
@vctt94 

> Right now it is possible to change vote option in a finished proposal.

Was fixed on my latest commit, could you please give it a try on your end.

File: app/components/views/ProposalDetails/ProposalDetails.jsx (hunk 5 above)
Comment (amassarwi) [reply app/components/views/ProposalDetails/ProposalDetails.jsx]: Done mate :mate:

**Comment by amassarwi** [comment]
@vctt94  

> I have found another issue which also happening on master, currently when voting on a proposal successfully the cast vote button is still enabled and isn't turning to disabled, I think we should fix both in this diff as well.

^ Addressed :heavy_check_mark:

**Comment by amassarwi** [comment]
@vctt94 
Just found another issue which I was able to reproduce on master that I would like to address here as well, `purchase tickets` button isn't navigating to the tickets page, see screenshot bellow:
image

File: app/style/Icons.less (hunk 6 above)
Comment (amassarwi) [reply app/style/Icons.less:6]: Looks like adding this change to this PR would make extraordinarily big, will create an issue for that and I would like to work on it after merging this one!

File: app/components/shared/ExternalLink.js (hunk 2 above)
Comment (amassarwi) [reply app/components/shared/ExternalLink.js]: Done, Instead of changing `ExternalLink` which is used in many other parts of the APP, I have refactored `PoliteiaLink` to not rely on `useExternalLink`, and to use combination of `pi-ui`'s `Link` & `Component` depends on the usage.

File: app/components/shared/ExternalLink.js [hunk 11]
```diff
@@ -2,6 +2,7 @@ import { shell } from "electron";
 import { network } from "connectors";
 
 const clicker = (isTestNet, href, hrefTestNet) => () => {
+  console.log(isTestNet);
```
Comment (amassarwi) [inline app/components/shared/ExternalLink.js:5]: remove log!

File: app/components/layout/TabbedPage.js [hunk 12]
```diff
@@ -116,8 +117,7 @@ function TabbedPage(props) {
       : "l2r";
     setDir(dir);
     setMatchedTab(matchedTab);
-  }, [ location ]);
-  let { children, header, className } = props;
+  }, [ location, previous, children, onChange ]);
```
Comment (vctt94) [inline app/components/layout/TabbedPage.js:120]: I removed previous from here and felt no side effects.
